src/coreclr/vm/interpexec.cpp

```diff
@@ -143,6 +144,142 @@ static OBJECTREF CreateMultiDimArray(MethodTable* arrayClass, int8_t* stack, int
+
+    // First, promote the source value to double
+    double src = LOCAL_VAR(ip[2], TSource),
+        minValue = (double)std::numeric_limits<TResult>::lowest(),
+        maxValue = (double)std::numeric_limits<TResult>::max();
+
+    // (src != src) checks for NaN, then we check whether the min and max values (as represented by their closest double)
+    //  properly bound the source value so that when it is truncated it will be in range
+    // We assume that we are in round-towards-zero mode. For NaN we want to return 0, and for out of range values, saturate.
+    TResult result;
+    if (src != src)
+        result = 0;
+    else if (src >= maxValue)
+        result = std::numeric_limits<TResult>::max();
+    else if (!std::numeric_limits<TResult>::is_signed && (src <= -1))
+        result = 0;
+    else if (std::numeric_limits<TResult>::is_signed && (src < minValue))
+        result = std::numeric_limits<TResult>::lowest();
+    else
+        result = (TResult)src;
+
+    // According to spec, for result types smaller than int32, we store them on the stack as int32
+    if (sizeof(TResult) >= 4)
+        LOCAL_VAR(ip[1], TResult) = result;
+    else
+        LOCAL_VAR(ip[1], int32_t) = (int32_t)result;
+}
+
+template <typename TResult, typename TSource> static void ConvOvfFpHelper(int8_t *stack, const int32_t *ip)
+{
+    static_assert(!std::numeric_limits<TSource>::is_integer, "ConvOvfFpHelper is only for use on floats and doubles");
+    static_assert(std::numeric_limits<TResult>::is_integer, "ConvOvfFpHelper is only for use on floats and doubles to be converted to integers");
+    static_assert(sizeof(TResult) <= 4, "ConvOvfFpHelper's generic version is only for use on results <= 4 bytes in size");
+
+    // First, promote the source value to double
+    double src = LOCAL_VAR(ip[2], TSource),
+        minValue = (double)std::numeric_limits<TResult>::lowest() - 1,
+        maxValue = (double)std::numeric_limits<TResult>::max() + 1;
+
+    // (src != src) checks for NaN, then we check whether the min and max values properly bound the source value
+    // We assume that we are in round-towards-zero mode
+    bool outOfRange = (src != src) || (src < minValue) || (src > maxValue);
+
+    if (outOfRange)
+        COMPlusThrow(kOverflowException);
+
+    TResult truncated = (TResult)src;
+    // According to spec, for result types smaller than int32, we store them on the stack as int32
+    LOCAL_VAR(ip[1], int32_t) = (int32_t)truncated;
+}
+
+// I64 and U64 versions based on mono_math.h
+
+template <typename TSource> static void ConvOvfFpHelperI64(int8_t *stack, const int32_t *ip)
+{
+    static_assert(!std::numeric_limits<TSource>::is_integer, "ConvOvfFpHelper is only for use on floats and doubles");
+
+	const double two63  = 2147483648.0 * 4294967296.0;
+    // First, promote the source value to double
+    double src = LOCAL_VAR(ip[2], TSource),
+        // Define the boundary values we need to be between (see mono_try_trunc_i64)
```

Comment:
```suggestion
        // Define the boundary values we need to be between (see System.Math.ConvertToInt64Checked)
```

Reply:
I don't think it is correct for the comment to point to something other than what the code is based on.

Reply:
We have the same code copied in multiple places. I think it is better to point to the managed implementation as the master implementation.